### PR TITLE
refactor(router/atc): do not create new table in `split_routes_and_services_by_path()`

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

The original `split_routes_and_services_by_path()` will create a new table to contains all routes even if no splitting.
To erase this overhead, now `split_routes_and_services_by_path()` will insert the cloned routes directly into the passed in table.

KAG-4138

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

**Comment by chronolaw** [comment]
@hanshuebner , could you take a look? since you are the author of this feature.

**Review by hanshuebner — COMMENTED** [review]
What is the benefit of this change.  I find that the `split_routes_and_services_by_path` function was already pretty complicated, and the change further complicates it.  I would endorse this change only if the performance benefit is measureable and significant.

**Comment by chronolaw** [comment]
I did a simple test:
```
  it("#only test", function()
    local transform = require("kong.router.transform")
    local split_routes_and_services_by_path = transform.split_routes_and_services_by_path

    local routes = {}

    for i = 1, 5000 do
      routes[i] = {
        route = {
          id = "e8fb37f1-102d-461e-9c51-6608a6bb8101",
          hosts = { "a.com" },
        }
      }
    end

    local COUNT = 1000

    ngx.update_time()
    local t = ngx.now()

    for i = 1, COUNT do
      split_routes_and_services_by_path(routes)
    end

    ngx.update_time()
    print(ngx.now() - t)
  end)
```

The master's output:
```
(kong-dev) [KONG kong]$bin/busted t.lua -t only
0.03000020980835
●
1 success / 0 failures / 0 errors / 0 pending : 0.063314 seconds
(kong-dev) [KONG kong]$bin/busted t.lua -t only
0.028000116348267
●
1 success / 0 failures / 0 errors / 0 pending : 0.059201 seconds
```

This PR's output:
```
(kong-dev) [KONG kong]$bin/busted t.lua -t only
0.017999887466431
●
1 success / 0 failures / 0 errors / 0 pending : 0.049651 seconds
(kong-dev) [KONG kong]$bin/busted t.lua -t only
0.016999959945679
●
1 success / 0 failures / 0 errors / 0 pending : 0.053319 seconds
```
Obviously, This PR is faster. @hanshuebner , Do you have more opinions?

**Comment by hanshuebner** [comment]
I think the performance optimization is too visible in the code, making it harder to understand and follow.  It was already pretty difficult to understand.  Given the tiny benefit, I would not go ahead and merge it.

**Review by bungle — CHANGES_REQUESTED** [review]
@chronolaw, I guess you will get the idea?

File: kong/router/transform.lua [hunk 1]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
+      route.paths = paths
+      route.id = uuid_generator(original_route_id .. "#" .. tostring(idx))
 
-local function split_routes_and_services_by_path(routes_and_services)
-  local count = #routes_and_services
-  local routes_and_services_split = tb_new(count, 0)
+      local cloned_route_and_service = {
+        route = route,
+        service = original_service,
+      }
```
Comment (bungle) [inline kong/router/transform.lua:727]: If we can optimize the first case, then I would remove this
```suggestion
```

File: kong/router/transform.lua [hunk 2]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
-    split_route_by_path_info(routes_and_services[i], routes_and_services_split)
-  end
+      if is_first then
+        -- the first one will replace the original route
+        routes_and_services[i] = cloned_route_and_service
+        is_first = false
+
+      else
+        -- the others will append to the original routes array
+        routes_and_services[count + append_count] = cloned_route_and_service
```
Comment (bungle) [inline kong/router/transform.lua:736]: And just do it here:

```suggestion
        routes_and_services[count + append_count] = {
        route = route,
        service = original_service,
      }
```

File: kong/router/transform.lua [hunk 3]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
-  local grouped_paths = group_by(
-    original_route.paths, sort_by_regex_or_length
-  )
-  for index, paths in pairs(grouped_paths) do
-    local cloned_route = {
-      route = shallow_copy(original_route),
-      service = route_and_service.service,
-    }
+    -- make sure that route_and_service contains only the two expected entries, route and service
+    assert(tb_nkeys(route_and_service) == 1 or tb_nkeys(route_and_service) == 2)
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua:710]: done, thanks.

File: kong/router/transform.lua [hunk 4]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
-    cloned_route.route.id = uuid_generator(original_route.id .. "#" .. tostring(index))
+    local original_route_id = original_route.id
+    local original_service = route_and_service.service
+    local grouped_paths = group_by(original_paths, sort_by_regex_or_length)
 
-    tb_insert(routes_and_services_split, cloned_route)
-  end
-end
+    local is_first = true
+    for idx, paths in pairs(grouped_paths) do
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua:717]: No, it is a hash table, like {true = 'regex', 2='/a', 3='/aa'}

File: kong/router/transform.lua [hunk 5]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
-  end
+  for i = 1, count do
+    local route_and_service = routes_and_services[i]
+    local original_route = route_and_service.route
+    local original_paths = original_route.paths
 
-  -- make sure that route_and_service contains only the two expected entries, route and service
-  assert(tb_nkeys(route_and_service) == 1 or tb_nkeys(route_and_service) == 2)
+    if is_empty_field(original_paths) or #original_paths == 1 or
+       not is_null(original_route.expression) -- expression will ignore paths
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua:704]: Yes, now 'expressions' flavor supports mixed mode, it means route may contains `paths` field.

File: kong/router/transform.lua [hunk 6]
```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
...
-    cloned_route.route.paths = paths
-    cloned_route.route.id = uuid_generator(original_route.id .. "#" .. tostring(index))
+    local original_route_id = original_route.id
+    local original_service = route_and_service.service
+    local grouped_paths = group_by(original_paths, sort_by_regex_or_length)
 
-    tb_insert(routes_and_services_split, cloned_route)
-  end
-end
+    local is_first = true
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua:716]: `grouped_paths` let the code  a little confusing, it is a hash table ,key is boolean or integer.

File: kong/router/transform.lua (hunk 1 above)
Comment (chronolaw) [reply kong/router/transform.lua:727]: As we talked above, we must clone the original route_and_service, and set the new paths and id.

File: kong/router/transform.lua (hunk 2 above)
Comment (chronolaw) [reply kong/router/transform.lua:740]: As Hans said, this logic is a little hard to understand, I also spent many time to figure it out then refactor it.

File: kong/router/transform.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua]: I renamed `idx` to `key` to make the code more clear.

File: kong/router/transform.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/router/transform.lua:704]: @chronolaw, ok, thanks for confirming.

File: kong/router/transform.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/router/transform.lua]: Can we change from Hash table to array? What is bennefit of `true`=`regex`?

**Comment by chronolaw** [comment]
> @chronolaw here are some of my proposals, feel free to merge them here or ignore: #12901

Thank you Aapo, it is a good improvement.
But I think it also increase the complexity of this PR, could you turn it into a new PR after this PR is merged?

File: kong/router/transform.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/transform.lua]: I try my best to keep the original code, do not change too much.

**Comment by bungle** [comment]
> > @chronolaw here are some of my proposals, feel free to merge them here or ignore: #12901
> 
> Thank you Aapo, it is a good improvement. But I think it also increase the complexity of this PR, could you turn it into a new PR after this PR is merged?

Ok.

**Review by bungle — APPROVED** [review]

**Pull request merged**